llvm/test/SYCLLowerIR/ESIMD/lower_invoke_simd.ll

```diff
@@ -56,8 +59,18 @@ define dso_local spir_func noundef float @SPMD_CALLER(float noundef %x, <16 x fl
   %res3 = call spir_func float @_Z33__regcall3____builtin_invoke_simdXX(<16 x float> (<16 x float> (<16 x float>)*, <16 x float>)* @SIMD_CALL_HELPER, <16 x float> (<16 x float>)* %raw_fptr,  float %x)
 ; CHECK: %{{.*}} = call spir_func float @_Z33__regcall3____builtin_invoke_simdXX(<16 x float> (<16 x float> (<16 x float>)*, <16 x float>)* @SIMD_CALL_HELPER, <16 x float> (<16 x float>)* %{{.*}}, float %{{.*}})
 
-  %res4 = fadd float %res1, %res2
-  %res = fadd float %res3, %res4
+;---- The invoke_simd calls.
+; Test case when function pointer is passed the __builtin_invoke_simd
+; and the actual function can be deduced.
+  %res4 = call spir_func float @_Z33__regcall3____builtin_invoke_simdYY(<16 x float> (<16 x float> (<16 x float>)* addrspace(4)*, <16 x float>)* noundef nonnull @SIMD_CALL_HELPER_YY, <16 x float> (<16 x float>)* noundef nonnull @SIMD_CALLEE_YY, float noundef 0.000000e+00)
```

Comment:
I think the test case has a bug. The second actual argument of the `invoke_simd` - `@SIMD_CALLEE_YY` - is a function pointer. It becomes the first formal argument of `@SIMD_CALL_HELPER_YY` - `<16 x float> (<16 x float>)* addrspace(4)* noundef align 8 dereferenceable(8) %f`, which is a pointer to function pointer.

Reply:
That LLVM IR came from this test case below. There may be some error in IR generation, I am re-checking that again,
but do you agree that the test itself is correct?

```#include <CL/sycl.hpp>
#include <sycl/ext/intel/esimd.hpp>
#include <sycl/ext/oneapi/experimental/invoke_simd.hpp>
using namespace sycl::ext::oneapi::experimental;
namespace esimd = sycl::ext::intel::esimd;
constexpr int VL = 16;
typedef simd<float, VL> __regcall (*FuncType)(simd<float, VL>);

[[intel::device_indirectly_callable]] SYCL_EXTERNAL simd<float, VL> __regcall intersectFunc(simd<float, VL> arg) SYCL_ESIMD_FUNCTION;// {return arg;}

SYCL_EXTERNAL void testfn() {
    auto sg = this_sub_group();
    float test = 0;
    FuncType func = intersectFunc;
    float res = invoke_simd(sg, func, test);
}
```

Reply:
'pointer to function pointer' came from this 'else' branch here (see that a pointer to 'f', which is a pointer to a function is passed):

```
  if constexpr (is_function) {
    return __builtin_invoke_simd<true /*function*/, RetSpmd>(
        detail::simd_func_call_helper<N, Callable, T...>, f,
        detail::unwrap_uniform<T>::impl(args)...);
  } else {
    // TODO support functors and lambdas which are handled in this branch.
    // The limiting factor for now is that the LLVMIR data flow analysis
    // implemented in LowerInvokeSimd.cpp which, finds actual invoke_simd
    // target function, can't handle this case yet.
    return __builtin_invoke_simd<false /*functor/lambda*/, RetSpmd>(
        detail::simd_obj_call_helper<N, Callable, T...>, &f,
        detail::unwrap_uniform<T>::impl(args)...);
  }
```